File: server/pkg/featureflag/static_provider.go
package featureflag

import (
	"context"
	"slices"
	"sync"
)

// Rule describes how a single flag is evaluated by the StaticProvider.
// All fields are optional; an empty Rule evaluates to Default (false) for
// everyone.
//
// Evaluation order (first match wins):
//
//  1. Deny: if any value in the EvalContext matches an entry in Deny on
//     attribute DenyBy (default "user_id"), the flag is OFF.
//  2. Allow: if any value matches an entry in Allow on attribute AllowBy
//     (default "user_id"), the flag is ON.
//  3. Percent: if Percent is non-nil and the bucket for (key, identifier)
//     falls inside Percent.Percent, the flag is ON.
//  4. Default: returned otherwise.
//
// Allow / Deny lists are intentionally separate (rather than a single
// targeting predicate) because operationally they cover different use
// cases — Allow is "internal users only" and Deny is "kill switch for
// these tenants" — and keeping them separate makes the data easy to audit
// in source control.
type Rule struct {
	// Default is the value returned when no targeting rule matches.
	Default bool

	// Variant is the variant identifier returned WHEN the rule evaluates
	// to enabled=true. For multi-arm experiments, set Variant to the
	// experiment-arm identifier (e.g. "experiment-v2"); for plain on/off
	// flags leave it empty.
	//
	// When the rule evaluates to enabled=false (default-off, deny hit,
	// percent miss, ...) the resulting Decision's Variant is always the
	// canonical "off". This is deliberate: a caller that branches on
	// Variant("checkout_algo", "control") would otherwise be routed into
	// the experiment arm even though the user did not roll into the
	// experiment cohort.
	Variant string

	// Allow is the set of identifier values that force the flag ON.
	Allow []string

	// AllowBy is the EvalContext attribute name used for Allow lookups.
	// Defaults to "user_id" when empty.
	AllowBy string

	// Deny is the set of identifier values that force the flag OFF.
	// Deny wins over Allow.
	Deny []string

	// DenyBy is the EvalContext attribute name used for Deny lookups.
	// Defaults to "user_id" when empty.
	DenyBy string

	// Percent enables a deterministic percent rollout. When nil, no
	// percent rollout is applied and Default is used as the fallback.
	Percent *PercentRollout
}

// PercentRollout describes a deterministic percent rollout.
//
// The bucket is computed from (flag key, EvalContext attribute By) using
// FNV-1a, which guarantees that the same identifier always falls into the
// same bucket across processes and across restarts. This is what callers
// need so users do not flip in and out of an experiment between requests.
type PercentRollout struct {
	// Percent is the rollout size in [0, 100]. 0 disables the rollout;
	// 100 enables it for everyone. Out-of-range values are clamped.
	Percent int

	// By selects the EvalContext attribute used as the bucketing
	// identifier. Defaults to "user_id". Use "workspace_id" for
	// workspace-scoped rollouts.
	By string
}

// StaticProvider is a thread-safe in-memory Provider populated either
// programmatically or from a config file. It is the recommended baseline
// provider for production: configuration lives in source control, moves
// through CD alongside the binary, and changes require a deploy — which is
// exactly the Continuous Delivery posture Martin Fowler recommends for
// Release Toggles and most Permissioning Toggles.
//
// For dynamic flags (kill switches, A/B tests changed by product) compose
// a StaticProvider with a DB-backed Provider behind a ChainProvider.
type StaticProvider struct {
	mu    sync.RWMutex
	rules map[string]Rule
}

// NewStaticProvider returns an empty StaticProvider. Use Set or
// LoadRules to populate it.
func NewStaticProvider() *StaticProvider {
	return &StaticProvider{rules: map[string]Rule{}}
}

// Name implements Provider.
func (*StaticProvider) Name() string { return "static" }

// Set installs or replaces the rule for key. Concurrent callers are
// serialized; readers (Lookup) never block writers for long.
func (p *StaticProvider) Set(key string, rule Rule) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.rules[key] = rule
}

// LoadRules atomically replaces every rule in the provider with the supplied
// map. Use this when reloading from a config file: a partial reload could
// otherwise leave the provider in a mixed state where some flags reflect the
// new config and others the old.
func (p *StaticProvider) LoadRules(rules map[string]Rule) {
	clone := make(map[string]Rule, len(rules))
	for k, v := range rules {
		clone[k] = v
	}
	p.mu.Lock()
	defer p.mu.Unlock()
	p.rules = clone
}

// Keys returns the sorted set of flag keys this provider knows about. Useful
// for diagnostic endpoints. The returned slice is a copy; mutating it does
// not affect the provider.
func (p *StaticProvider) Keys() []string {
	p.mu.RLock()
	defer p.mu.RUnlock()
	out := make([]string, 0, len(p.rules))
	for k := range p.rules {
		out = append(out, k)
	}
	slices.Sort(out)
	return out
}

// Lookup implements Provider.
func (p *StaticProvider) Lookup(ctx context.Context, key string) (Decision, bool) {
	p.mu.RLock()
	rule, ok := p.rules[key]
	p.mu.RUnlock()
	if !ok {
		return Decision{}, false
	}
	ec := EvalContextFrom(ctx)
	return evaluateRule(key, rule, ec), true
}

func evaluateRule(key string, rule Rule, ec EvalContext) Decision {
	// Deny wins over everything else. A kill switch must be reachable
	// even when other targeting matches.
	denyBy := orDefault(rule.DenyBy, "user_id")
	if len(rule.Deny) > 0 {
		if v, ok := ec.Lookup(denyBy); ok && slices.Contains(rule.Deny, v) {
			return decisionFromRule(key, rule, false, ReasonStatic)
		}
	}

	allowBy := orDefault(rule.AllowBy, "user_id")
	if len(rule.Allow) > 0 {
		if v, ok := ec.Lookup(allowBy); ok && slices.Contains(rule.Allow, v) {
			return decisionFromRule(key, rule, true, ReasonStatic)
		}
	}

	if rule.Percent != nil {
		by := orDefault(rule.Percent.By, "user_id")
		identifier, _ := ec.Lookup(by)
		// An empty identifier still produces a deterministic bucket
		// (the empty string hashes to a stable bucket) but in practice
		// that means everyone-without-an-id lands in the same bucket.
		// That's the desired behavior for percent rollouts at the edge:
		// anonymous users get a single shared rollout decision per
		// flag, not a uniformly random one.
		if inPercent(key, identifier, rule.Percent.Percent) {
			return decisionFromRule(key, rule, true, ReasonPercent)
		}
		return decisionFromRule(key, rule, false, ReasonPercent)
	}

	return decisionFromRule(key, rule, rule.Default, ReasonStatic)
}

func decisionFromRule(key string, rule Rule, enabled bool, reason Reason) Decision {
	// Variant policy: rule.Variant is the ON-variant. When the rule
	// evaluates to false we return the canonical "off" so a caller
	// branching on Variant() cannot accidentally enter the experiment
	// arm for a user that did not roll in.
	variant := boolToVariant(enabled)
	if enabled && rule.Variant != "" {
		variant = rule.Variant
	}
	return Decision{
		Key:     key,
		Enabled: enabled,
		Variant: variant,
		Reason:  reason,
		Source:  "static",
	}
}

func orDefault(v, def string) string {
	if v == "" {
		return def
	}
	return v
}

File: packages/core/feature-flags/static-provider.ts
import type { Decision, EvalContext, Provider, Rule } from "./types";
import { inPercent } from "./hash";

/**
 * StaticProvider is an in-memory Provider populated either programmatically
 * or from a JSON config shipped with the application bundle.
 *
 * This is the recommended baseline provider for the frontend: configuration
 * lives in source control, moves through CD alongside the build, and
 * changes require a deploy. For dynamic flags fetched from the backend,
 * wrap a {@link StaticProvider} behind a chain provider that also reads
 * from API state — the StaticProvider then acts as a safety net for the
 * very first paint before the API response is available.
 */
export class StaticProvider implements Provider {
  readonly name = "static";
  private rules: Map<string, Rule>;

  constructor(rules: Readonly<Record<string, Rule>> = {}) {
    this.rules = new Map(Object.entries(rules));
  }

  /** Replace or install the rule for `key`. */
  set(key: string, rule: Rule): void {
    this.rules.set(key, rule);
  }

  /**
   * Replace every rule atomically. Use when reloading flag config from a
   * fetch response so consumers never observe a mixed state.
   */
  loadRules(rules: Readonly<Record<string, Rule>>): void {
    this.rules = new Map(Object.entries(rules));
  }

  /** Sorted list of known flag keys. Useful for dev overlays. */
  keys(): string[] {
    return Array.from(this.rules.keys()).sort();
  }

  lookup(key: string, ctx: EvalContext): Decision | undefined {
    const rule = this.rules.get(key);
    if (!rule) return undefined;
    return evaluateRule(key, rule, ctx);
  }
}

function evaluateRule(key: string, rule: Rule, ctx: EvalContext): Decision {
  // Deny wins over everything else; a kill switch must remain reachable
  // even when other targeting matches.
  const denyBy = rule.denyBy ?? "user_id";
  if (rule.deny && rule.deny.length > 0) {
    const v = lookupAttr(ctx, denyBy);
    if (v && rule.deny.includes(v)) {
      return decisionFromRule(key, rule, false, "static");
    }
  }

  const allowBy = rule.allowBy ?? "user_id";
  if (rule.allow && rule.allow.length > 0) {
    const v = lookupAttr(ctx, allowBy);
    if (v && rule.allow.includes(v)) {
      return decisionFromRule(key, rule, true, "static");
    }
  }

  if (rule.percent) {
    const by = rule.percent.by ?? "user_id";
    const ident = lookupAttr(ctx, by) ?? "";
    const enabled = inPercent(key, ident, rule.percent.percent);
    return decisionFromRule(key, rule, enabled, "percent");
  }

  return decisionFromRule(key, rule, rule.default ?? false, "static");
}

function decisionFromRule(
  key: string,
  rule: Rule,
  enabled: boolean,
  reason: Decision["reason"],
): Decision {
  // Variant policy: rule.variant is the ON-variant. When the rule
  // evaluates to false we return the canonical "off" so a caller
  // branching on the variant cannot accidentally enter the experiment
  // arm for a user that did not roll in.
  let variant = boolToVariant(enabled);
  if (enabled && rule.variant && rule.variant.length > 0) {
    variant = rule.variant;
  }
  return {
    key,
    enabled,
    variant,
    reason,
    source: "static",
  };
}

function boolToVariant(b: boolean): string {
  return b ? "on" : "off";
}

/**
 * Resolve an attribute name against the EvalContext. The well-known names
 * "user_id" and "workspace_id" map to the dedicated fields so rules can use
 * them by name without callers also populating `attributes`.
 */
function lookupAttr(ctx: EvalContext, name: string): string | undefined {
  if (name === "user_id") return nonEmpty(ctx.userId);
  if (name === "workspace_id") return nonEmpty(ctx.workspaceId);
  return nonEmpty(ctx.attributes?.[name]);
}

function nonEmpty(v: string | undefined): string | undefined {
  return v && v.length > 0 ? v : undefined;
}
